data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/common/utils/CaffeineCache.java

```diff
@@ -0,0 +1,131 @@
+    private long curBytes;
+    private boolean strictLRU;
+
+    public CaffeineCache(int maxEntries, long maxBytes, int ttlSeconds, boolean strictLRU) {
+        this.maxBytes = maxBytes;
+        this.strictLRU = strictLRU;
+        this.maxEntries = maxEntries;
+        Caffeine<K, V> caffeine;
+        if (strictLRU) {
+            caffeine = (Caffeine<K, V>) Caffeine.newBuilder()
+                .maximumSize(maxEntries);
+        } else {
+            caffeine = (Caffeine<K, V>) Caffeine.newBuilder()
+                .maximumWeight(maxBytes)
+                .weigher(new ByteArrayValueWeigher());
+        }
+        this.cache = caffeine
+                .expireAfterAccess(ttlSeconds, TimeUnit.SECONDS)
+                .removalListener((K k, V v, RemovalCause cause) -> {
+                    int valueLength = (v instanceof String) ? ((String)v).length() : 
+                                ((Event)v).toJsonString().length();
+                    curBytes -= getEntrySize(k, v);
+                    curEntries--;
+                })
+                .recordStats()
+                .build();
+    }
+
+    long getEntrySize(K key, V value) {
+        long size = OVERHEAD_PER_CACHE_ENTRY;
+        if (key instanceof String) {
+            size += (long)((String)key).length();
+        } else {
+            size += 8L;
+        }
+        size += (value instanceof String) ? ((String)value).length() : 
+                    ((Event)value).toJsonString().length();
+        return size;
+    }
+
+    public void put(K key, V value) {
+        if (!(value instanceof Event) && !(value instanceof String)) {
+            throw new RuntimeException("Currently only Event/String type values are supported");
+        }
+        if (!(key instanceof String) && !(key instanceof Integer) && !(key instanceof Long)) {
+            throw new RuntimeException("Currently only String/Integer/Long type keys are supported");
+        }
+        cache.put(key, value);
+        curEntries++;
+        curBytes += getEntrySize(key, value);
+        // run cleanUp to create space by removing any expired entries
+        if (curEntries > maxEntries/2 || curBytes > maxBytes/2) {
+            cache.cleanUp();
+        }
+    }
+
+    public V get(K key) {
+        return cache.getIfPresent(key);
+    }
+
+    public boolean containsKey(K key) {
+        return cache.getIfPresent(key) != null;
+    }
+
+    public void remove(K key) {
+        V value = cache.getIfPresent(key);
+        if (value == null) {
+            return;
+        }
+        cache.invalidate(key);
+        curEntries--;
+        curBytes += getEntrySize(key, value);
+    }
+
+    public CacheStats getStats() {
+        return cache.stats();
+    }
+
+    public void clear() {
+        cache.invalidateAll();
+        cache.cleanUp();
+    }
+
+    @VisibleForTesting
+    int getNumEntries() {
+        return curEntries;
+    }
+
+    @VisibleForTesting
+    long getSize() {
+        return curBytes;
```

Comment:
fwiw, you can get the `weightedSize()` from `cache.policy().eviction()` if `isWeighted()`.